performance_manager/lib/l0_rt_vehicle_positions.py

```diff
@@ -160,275 +188,110 @@ def join_gtfs_static(
     )
-
-    with sql_session.begin() as session:
-        # Join records from db with records from parquet db records should
-        # contain 'index' column that does not exist in parquet dataset
-        return pandas.concat(
-            [
-                pandas.DataFrame(
-                    [row._asdict() for row in session.execute(get_db_events)]
-                ),
-                new_events,
-            ]
-        ).sort_values(by=["vehicle_id", "timestamp_start"])
-
-
-def merge_vehicle_position_events(
-    folder: str,
-    new_events: pandas.DataFrame,
-    session: sessionmaker,
-) -> None:
-    """
-    merge new events from parquet file with exiting events found in database
-    new events will be merged with existing events in a 5 minutes window
-    surrounding the year/month/day/hour value found in path of parquet files
-    """
-    process_logger = ProcessLogger("vp_merge_events")
     process_logger.log_start()
 
-    merge_events = get_event_overlap(folder, new_events, session)
-
-    # pylint: disable=duplicate-code
-
-    # Identify records that are continuing from existing db
-    # If such records are found, update timestamp_end with latest value
-    first_of_consecutive_events = merge_events["hash"] == merge_events[
-        "hash"
-    ].shift(-1)
-    last_of_consecutive_events = merge_events["hash"] == merge_events[
-        "hash"
-    ].shift(1)
-
-    merge_events["timestamp_end"] = numpy.where(
-        first_of_consecutive_events,
-        merge_events["timestamp_end"].shift(-1),
-        merge_events["timestamp_end"],
-    )
-
-    existing_was_updated_mask = (
-        ~(merge_events["pk_id"].isna()) & first_of_consecutive_events
+    vehicle_positions = add_event_hash_column(
+        vehicle_positions,
+        hash_column_name="trip_stop_hash",
+        expected_hash_columns=[
+            "stop_sequence",
+            "parent_station",
+            "direction_id",
+            "route_id",
+            "start_date",
+            "start_time",
+            "vehicle_id",
+        ],
     )
 
-    existing_to_del_mask = (
-        ~(merge_events["pk_id"].isna()) & last_of_consecutive_events
+    # create a pivot table on the trip stop hash, finding the earliest time
+    # that each vehicle/stop pair is and is not moving. rename the vehicle
+    # timestamps to vp_stop_timestamp and vp_move_timestamp, the names used
+    # in the database
+    vp_timestamps = pandas.pivot_table(
+        vehicle_positions,
+        index="trip_stop_hash",
+        columns="is_moving",
+        aggfunc={"vehicle_timestamp": min},
+    ).reset_index(drop=False)
+
+    vp_timestamps.columns = vp_timestamps.columns.to_flat_index()
+    vp_timestamps = vp_timestamps.rename(
+        columns={
+            ("trip_stop_hash", ""): "trip_stop_hash",
+            ("vehicle_timestamp", False): "vp_stop_timestamp",
+            ("vehicle_timestamp", True): "vp_move_timestamp",
+        }
```

Comment:
this logic matches what @rymarczy had in his branch. i really like the idea of using a pivot table, which reduces the numpy.where usage.